lib/MailNotifications.php

```diff
@@ -0,0 +1,412 @@
+		$notificationsCount = count($notifications);
+		$template->setSubject($l10n->n('New notification for %s', '%n new notifications for %s', $notificationsCount, [$this->defaults->getName()]));
+		$template->addBodyText(
+			$l10n->n('You have a new notification for %s', 'You have %n new notifications for %s', $notificationsCount, [$homeLink]),
+			$l10n->n('You have a new notification for %s', 'You have %n new notifications for %s', $notificationsCount, [$this->urlGenerator->getAbsoluteURL('/')])
+		);
+
+		// Prepare email body with the content of missed notifications
+		// Notifications are assumed to be passed in in descending order (latest first). Reversing to present chronologically.
+		$notifications = array_reverse($notifications);
+
+		$timezone = $this->config->getUserValue($uid, 'core', 'timezone', date_default_timezone_get());
+
+		foreach ($notifications as $notification) {
+			$relativeDateTime = $this->dateFormatter->formatDateTimeRelativeDay($notification->getDateTime(), 'long', 'short', new \DateTimeZone($timezone), $l10n);
+			$template->addBodyListItem($this->getHTMLContents($notification), $relativeDateTime, $notification->getIcon(), $notification->getParsedSubject());
+
+			// Buttons probably were not intended for this, but it works ok enough for showing the idea.
+			$actions = $notification->getParsedActions();
+			foreach ($actions as $action) {
+				$template->addBodyButton($action->getLabel(), $action->getLink());
+			}
+		}
+
+		// Prepare email footer
+		$template->addBodyText(
+			$l10n->t('You can change the frequency of these emails or disable them <a href="%s">here</a>.', $this->urlGenerator->getAbsoluteURL('/settings/user/notifications')),
+			$l10n->t('You can change the frequency of these emails or disable them here: %s', $this->urlGenerator->getAbsoluteURL('/settings/user/notifications'))
+		);
+
+		$template->addFooter();
+
+		$message = $this->mailer->createMessage();
+		$message->useTemplate($template);
+		$message->setTo([$userEmailAddress => $user->getDisplayName()]);
+		$message->setFrom([Util::getDefaultEmailAddress('no-reply') => $this->defaults->getName()]);
+
+		return $message;
+	}
+
+	/**
+	 * return HTML to display this notification
+	 *
+	 * @param INotification $notification
+	 * @return string
+	 */
+	protected function getHTMLContents(INotification $notification): string {
+		$HTMLSubject = $this->getHTMLSubject($notification);
+		$link = $notification->getLink();
+		if ($link !== '') {
+			$HTMLSubject = '<a href="' . $link . '">' . $HTMLSubject . '</a>';
+		}
+
+		return $HTMLSubject . '<br>' . $this->getHTMLMessage($notification);
+	}
+
+	/**
+	 * return HTML to display the subject of this notification
+	 *
+	 * @param INotification $notification
+	 * @return string
+	 */
+	protected function getHTMLSubject(INotification $notification): string {
+		$contentString = $notification->getRichSubject();
```

Comment:
Should use getParsedSubject, which is HTML save (should be plain text only).
Otherwise you will get weird emails in most cases of open security issues

Reply:
Ah, I see you replace the rich parameters later, then I would say just `htmlspecialchars` the `$contentString` as well

Reply:
Added htmlspecialchars in both getHTMLSubject and getHTMLMessage